wallet/src/key_chain/leaf_key_chain.rs

```diff
@@ -0,0 +1,517 @@
+        derive_if_necessary: bool,
+    ) -> bool {
+        // The public_key derivation path must be longer than the parent of this key chain
+        if let Some(path_diff) = public_key
+            .get_derivation_path()
+            .get_super_path_diff(self.parent_pubkey.get_derivation_path())
+        {
+            // The path difference must be 1 i.e. the key index
+            if path_diff.len() == 1 {
+                let key_index = path_diff[0];
+                // Check if we expect this key to be derived
+                let is_pub_key_derived = match self.get_last_derived_index() {
+                    None => false,
+                    Some(last_derived_index) => {
+                        key_index.get_index() <= last_derived_index.get_index()
+                    }
+                };
+
+                if is_pub_key_derived {
+                    if let Some(pk) = self.derived_public_keys.get(&key_index) {
+                        return pk == public_key;
+                    }
+                } else if derive_if_necessary {
+                    if let Ok(pk) = &self.derive_key(key_index) {
+                        return pk == public_key;
+                    }
+                }
+            }
+        }
+        false
+    }
+
+    pub fn is_public_key_mine(&self, public_key: &PublicKey) -> bool {
+        self.public_key_to_index.contains_key(public_key)
+    }
+
+    pub fn is_public_key_hash_mine(&self, pubkey_hash: &PublicKeyHash) -> bool {
+        self.public_key_hash_to_index.contains_key(pubkey_hash)
+    }
+
+    /// Get the extended public key provided a public key or None if no key found
+    pub fn get_xpub_from_public_key(&self, pub_key: &PublicKey) -> Option<&ExtendedPublicKey> {
+        self.derived_public_keys.get(self.public_key_to_index.get(pub_key)?)
+    }
+
+    /// Get the extended public key provided a public key hash or None if no key found
+    pub fn get_xpub_from_public_key_hash(&self, pkh: &PublicKeyHash) -> Option<&ExtendedPublicKey> {
+        self.derived_public_keys.get(self.public_key_hash_to_index.get(pkh)?)
+    }
+
+    /// Mark a specific key as used in the key pool. This will update the last used key index if
+    /// necessary. Returns false if a key was found and set to used.
+    pub fn mark_extended_pubkey_as_used<B: Backend>(
+        &mut self,
+        db_tx: &mut StoreTxRw<B>,
+        extended_public_key: &ExtendedPublicKey,
+        lookahead_size: u32,
+    ) -> KeyChainResult<bool> {
+        // Check if public key is in the key pool
+        if self.is_pubkey_mine_in_key_pool(extended_public_key) {
+            // Get the key index of the public key, this should always be Some
+            let key_index = extended_public_key.get_derivation_path().as_slice().last()
+                .expect("The provided public key belongs to this key chain, hence it should always have a key index");
+            self.usage_state.set_last_used(Some(*key_index));
```

Comment:
This may override the last used to a lower value, which is probably not a good idea.

Reply:
This is actually one of the reasons I have been against having explicit way to set_last_used instead of something that increments + checks based on the internal state.